tests/suites/test_suite_mpi.function

```diff
@@ -1720,6 +1720,406 @@ exit:
 }
 /* END_CASE */
 
+/* BEGIN_CASE */
+void mpi_core_add_if( char * input_A, char * input_B,
+                      char * input_X4, int carry4,
+                      char * input_X8, int carry8 )
+{
+    mbedtls_mpi X4, X8, A, B;
+    mbedtls_mpi_uint *a = NULL; /* first value to add */
+    mbedtls_mpi_uint *b = NULL; /* second value to add */
+    mbedtls_mpi_uint *x = NULL; /* expected */
+    mbedtls_mpi_uint *d = NULL; /* destination - the in/out first op */
+
+    mbedtls_mpi_init( &A );
+    mbedtls_mpi_init( &B );
+    mbedtls_mpi_init( &X4 );
+    mbedtls_mpi_init( &X8 );
+
+    TEST_EQUAL( 0, mbedtls_test_read_mpi( &A, input_A ) );
+    TEST_EQUAL( 0, mbedtls_test_read_mpi( &B, input_B ) );
+    TEST_EQUAL( 0, mbedtls_test_read_mpi( &X4, input_X4 ) );
+    TEST_EQUAL( 0, mbedtls_test_read_mpi( &X8, input_X8 ) );
+
+    /* We only need to work with one of (X4, carry4) or (X8, carry8) depending
+     * on sizeof(mbedtls_mpi_uint)
+     */
+    mbedtls_mpi *X = ( sizeof(mbedtls_mpi_uint) == 4 ) ? &X4 : &X8;
+    mbedtls_mpi_uint carry = ( sizeof(mbedtls_mpi_uint) == 4 ) ? carry4 : carry8;
+
+    /* All of the inputs are +ve (or zero) */
+    TEST_EQUAL( 1, A.s );
+    TEST_EQUAL( 1, B.s );
+    TEST_EQUAL( 1, X->s );
+
+    /* Test cases are such that A <= B, so #limbs should be <= */
+    TEST_ASSERT( A.n <= B.n );
+    TEST_ASSERT( X->n <= B.n );
+
+    /* Now let's get arrays of mbedtls_mpi_uints, rather than MPI structures */
+
+    /* mbedtls_mpi_core_add_if() uses input arrays of mbedtls_mpi_uints which
+     * must be the same size. The MPIs we've read in will only have arrays
+     * large enough for the number they represent. Therefore we create new
+     * raw arrays of mbedtls_mpi_uints and populate them from the MPIs we've
+     * just read in.
+     *
+     * We generated test data such that B was always >= A, so that's how many
+     * limbs each of these need.
+     */
+    size_t limbs = B.n;
+    a = mbedtls_calloc( limbs, sizeof(mbedtls_mpi_uint) );
+    b = mbedtls_calloc( limbs, sizeof(mbedtls_mpi_uint) );
+    x = mbedtls_calloc( limbs, sizeof(mbedtls_mpi_uint) );
+    d = mbedtls_calloc( limbs, sizeof(mbedtls_mpi_uint) );
+
+    size_t bytes = limbs * sizeof(mbedtls_mpi_uint);
+
+    TEST_ASSERT( a != NULL );
+    TEST_ASSERT( b != NULL );
+    TEST_ASSERT( x != NULL );
+    TEST_ASSERT( d != NULL );
```

Comment:
> It would be much better if the name contained CALLOC, to make it clear that memory is zero'd.

Well, maybe. `ASSERT_ALLOC` has the init-to-zero aspect of calloc, but not the multiply-by-element-size aspect. We don't even have a way to allocate non-zeroed memory in mbedtls (no `malloc` primitive in the platform abstracction), so any allocation has init-to-zero. I'd find `ASSERT_CALLOC` misleading if it didn't have two size arguments, and the element size is almost always 1 in our tests (bignum limbs being the main exception).